File: fastapi-backend/app/services/realtime_stt_service.py
"""
Professional RealtimeSTT Service for Real-time Speech Recognition
Replaces faster-whisper with RealtimeSTT for better real-time performance
"""

import asyncio
import logging
import threading
import time
from typing import Optional, Dict, Any, Callable, List
from dataclasses import dataclass
from RealtimeSTT import AudioToTextRecorder
from app.utils.logger import get_logger

logger = get_logger(__name__)

@dataclass
class RealtimeSTTConfig:
    """Configuration for RealtimeSTT service."""
    model: str = "base.en"  # Model size: tiny.en, base.en, small.en, medium.en
    language: str = "en"  # Language code
    use_microphone: bool = True  # Use microphone input
    post_speech_silence_duration: float = 1.0  # Silence duration to end speech
    spinner: bool = False  # Show loading spinner
    enable_realtime_transcription: bool = True  # Enable real-time updates
    enable_final_transcription: bool = True  # Enable final transcription
    sample_rate: int = 16000  # Audio sample rate
    chunk_size: int = 1024  # Audio chunk size

@dataclass
class RealtimeSTTCallbacks:
    """Callbacks for RealtimeSTT events."""
    on_realtime_transcription: Optional[Callable[[str, float, bool], None]] = None
    on_final_transcription: Optional[Callable[[str, float], None]] = None
    on_error: Optional[Callable[[str], None]] = None
    on_state_change: Optional[Callable[[bool], None]] = None
    on_language_detected: Optional[Callable[[str], None]] = None

class RealtimeSTTService:
    """
    Professional RealtimeSTT service for real-time speech recognition.
    
    Features:
    - Real-time transcription with live updates
    - Final transcription when speech ends
    - Language detection and validation
    - Error handling and recovery
    - Thread-safe operations
    - Professional logging
    """
    
    def __init__(self, config: RealtimeSTTConfig, callbacks: RealtimeSTTCallbacks):
        self.config = config
        self.callbacks = callbacks
        self.recorder: Optional[AudioToTextRecorder] = None
        self.is_initialized = False
        self.is_recording = False
        self.is_connected = False
        self._lock = threading.Lock()
        self._recorder_thread: Optional[threading.Thread] = None
        self._stop_event = threading.Event()
        
        # Statistics
        self.total_transcriptions = 0
        self.successful_transcriptions = 0
        self.failed_transcriptions = 0
        self.start_time = None
        
        logger.info(f"RealtimeSTT Service initialized with model: {config.model}, language: {config.language}")

    async def initialize(self) -> bool:
        """
        Initialize the RealtimeSTT service.
        
        Returns:
            bool: True if initialization successful, False otherwise
        """
        try:
            logger.info(f"Initializing RealtimeSTT with model: {self.config.model}")
            
            # Initialize AudioToTextRecorder in a separate thread
            loop = asyncio.get_event_loop()
            self.recorder = await loop.run_in_executor(
                None,
                self._create_recorder
            )
            
            self.is_initialized = True
            self.is_connected = True
            self.start_time = time.time()
            
            logger.info(f"RealtimeSTT initialized successfully with model: {self.config.model}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to initialize RealtimeSTT: {e}")
            self.is_initialized = False
            self.is_connected = False
            return False

    def _create_recorder(self) -> AudioToTextRecorder:
        """Create and configure the AudioToTextRecorder."""
        try:
            # Configure recorder with callbacks
            recorder = AudioToTextRecorder(
                model=self.config.model,
                language=self.config.language,
                use_microphone=self.config.use_microphone,
                on_realtime_transcription_update=self._on_realtime_transcription,
                spinner=self.config.spinner,
                post_speech_silence_duration=self.config.post_speech_silence_duration
            )
            
            # Override the final transcription callback
            recorder._on_final_transcription = self._on_final_transcription
            
            return recorder
            
        except Exception as e:
            logger.error(f"Failed to create AudioToTextRecorder: {e}")
            raise

    def _on_realtime_transcription(self, text: str):
        """Handle real-time transcription updates."""
        try:
            if not text or not text.strip():
                return
                
            # Calculate confidence (simple heuristic)
            confidence = min(1.0, max(0.0, len(text.strip()) / 50.0))
            
            # Update statistics
            self.total_transcriptions += 1
            
            logger.debug(f"Realtime transcription: {text[:50]}...")
            
            # Call user callback
            if self.callbacks.on_realtime_transcription:
                try:
                    self.callbacks.on_realtime_transcription(text, confidence, False)
                except Exception as e:
                    logger.error(f"Error in realtime transcription callback: {e}")
                    
        except Exception as e:
            logger.error(f"Error in realtime transcription handler: {e}")
            self.failed_transcriptions += 1

    def _on_final_transcription(self, text: str):
        """Handle final transcription when speech ends."""
        try:
            if not text or not text.strip():
                return
                
            # Calculate confidence
            confidence = min(1.0, max(0.0, len(text.strip()) / 50.0))
            
            # Update statistics
            self.successful_transcriptions += 1
            
            logger.info(f"Final transcription: {text}")
            
            # Call user callback
            if self.callbacks.on_final_transcription:
                try:
                    self.callbacks.on_final_transcription(text, confidence)
                except Exception as e:
                    logger.error(f"Error in final transcription callback: {e}")
                    
        except Exception as e:
            logger.error(f"Error in final transcription handler: {e}")
            self.failed_transcriptions += 1

    async def start(self) -> bool:
        """
        Start the RealtimeSTT service.
        
        Returns:
            bool: True if started successfully, False otherwise
        """
        try:
            with self._lock:
                if not self.is_initialized:
                    logger.warning("RealtimeSTT not initialized, cannot start")
                    return False
                    
                if self.is_recording:
                    logger.warning("RealtimeSTT already recording")
                    return True
                
                logger.info("Starting RealtimeSTT service...")
                
                # Start recording in a separate thread
                self._stop_event.clear()
                self._recorder_thread = threading.Thread(
                    target=self._recording_loop,
                    daemon=True
                )
                self._recorder_thread.start()
                
                self.is_recording = True
                
                # Notify state change
                if self.callbacks.on_state_change:
                    try:
                        self.callbacks.on_state_change(True)
                    except Exception as e:
                        logger.error(f"Error in state change callback: {e}")
                
                logger.info("RealtimeSTT service started successfully")
                return True
                
        except Exception as e:
            logger.error(f"Failed to start RealtimeSTT service: {e}")
            return False

    def stop(self):
        """Stop the RealtimeSTT service."""
        try:
            with self._lock:
                if not self.is_recording:
                    logger.warning("RealtimeSTT not recording, cannot stop")
                    return
                
                logger.info("Stopping RealtimeSTT service...")
                
                # Signal stop
                self._stop_event.set()
                
                # Wait for thread to finish
                if self._recorder_thread and self._recorder_thread.is_alive():
                    self._recorder_thread.join(timeout=2.0)
                
                self.is_recording = False
                
                # Notify state change
                if self.callbacks.on_state_change:
                    try:
                        self.callbacks.on_state_change(False)
                    except Exception as e:
                        logger.error(f"Error in state change callback: {e}")
                
                logger.info("RealtimeSTT service stopped")
                
        except Exception as e:
            logger.error(f"Error stopping RealtimeSTT service: {e}")

    def _recording_loop(self):
        """Main recording loop running in a separate thread."""
        try:
            logger.info("RealtimeSTT recording loop started")
            
            while not self._stop_event.is_set():
                try:
                    if self.recorder:
                        # This is a blocking call that waits for speech
                        # and calls the callbacks when speech is detected
                        self.recorder.text(self._on_final_transcription)
                    else:
                        logger.warning("Recorder not available, stopping loop")
                        break
                        
                except Exception as e:
                    logger.error(f"Error in recording loop: {e}")
                    if self.callbacks.on_error:
                        try:
                            self.callbacks.on_error(str(e))
                        except Exception as callback_error:
                            logger.error(f"Error in error callback: {callback_error}")
                    
                    # Wait a bit before retrying
                    time.sleep(0.1)
                    
        except Exception as e:
            logger.error(f"Fatal error in recording loop: {e}")
        finally:
            logger.info("RealtimeSTT recording loop ended")

    def update_config(self, new_config: RealtimeSTTConfig):
        """Update the configuration."""
        try:
            with self._lock:
                self.config = new_config
                logger.info("RealtimeSTT configuration updated")
        except Exception as e:
            logger.error(f"Error updating configuration: {e}")

    def get_status(self) -> Dict[str, Any]:
        """Get the current status of the service."""
        try:
            with self._lock:
                uptime = time.time() - self.start_time if self.start_time else 0
                
                return {
                    "is_initialized": self.is_initialized,
                    "is_recording": self.is_recording,
                    "is_connected": self.is_connected,
                    "model": self.config.model,
                    "language": self.config.language,
                    "total_transcriptions": self.total_transcriptions,
                    "successful_transcriptions": self.successful_transcriptions,
                    "failed_transcriptions": self.failed_transcriptions,
                    "uptime_seconds": uptime,
                    "success_rate": (
                        self.successful_transcriptions / max(1, self.total_transcriptions) * 100
                    )
                }
        except Exception as e:
            logger.error(f"Error getting status: {e}")
            return {
                "is_initialized": False,
                "is_recording": False,
                "is_connected": False,
                "error": str(e)
            }

    def destroy(self):
        """Clean up resources."""
        try:
            logger.info("Destroying RealtimeSTT service...")
            
            # Stop recording
            self.stop()
            
            # Clean up recorder
            if self.recorder:
                try:
                    self.recorder.shutdown()
                except Exception as e:
                    logger.error(f"Error shutting down recorder: {e}")
                finally:
                    self.recorder = None
            
            self.is_initialized = False
            self.is_connected = False
            
            logger.info("RealtimeSTT service destroyed")
            
        except Exception as e:
            logger.error(f"Error destroying RealtimeSTT service: {e}")

# Global service instance
_realtime_stt_service: Optional[RealtimeSTTService] = None

def get_realtime_stt_service() -> RealtimeSTTService:
    """Get the global RealtimeSTT service instance."""
    global _realtime_stt_service
    if _realtime_stt_service is None:
        # Default configuration
        config = RealtimeSTTConfig(
            model="base.en",
            language="en",
            use_microphone=True,
            post_speech_silence_duration=1.0,
            spinner=False,
            enable_realtime_transcription=True,
            enable_final_transcription=True
        )
        
        # Default callbacks
        callbacks = RealtimeSTTCallbacks()
        
        _realtime_stt_service = RealtimeSTTService(config, callbacks)
    
    return _realtime_stt_service

def create_realtime_stt_service(config: RealtimeSTTConfig, callbacks: RealtimeSTTCallbacks) -> RealtimeSTTService:
    """Create a new RealtimeSTT service instance."""
    return RealtimeSTTService(config, callbacks)

File: web-app/services/RealtimeSTTService.ts
/**
 * Professional RealtimeSTT Service for Real-time Speech Recognition
 * Replaces faster-whisper with RealtimeSTT for better real-time performance
 */

export interface RealtimeSTTConfig {
  model: string;
  language: string;
  useMicrophone: boolean;
  postSpeechSilenceDuration: number;
  spinner: boolean;
  enableRealtimeTranscription: boolean;
  enableFinalTranscription: boolean;
  sampleRate: number;
  chunkSize: number;
}

export interface RealtimeSTTCallbacks {
  onRealtimeTranscription?: (text: string, confidence: number, isFinal: boolean) => void;
  onFinalTranscription?: (text: string, confidence: number) => void;
  onError?: (error: string) => void;
  onStateChange?: (isRecording: boolean) => void;
  onLanguageDetected?: (language: string) => void;
}

export interface RealtimeSTTStatus {
  isInitialized: boolean;
  isRecording: boolean;
  isConnected: boolean;
  model: string;
  language: string;
  totalTranscriptions: number;
  successfulTranscriptions: number;
  failedTranscriptions: number;
  uptimeSeconds: number;
  successRate: number;
}

export class RealtimeSTTService {
  private ws: WebSocket | null = null;
  private config: RealtimeSTTConfig;
  private callbacks: RealtimeSTTCallbacks;
  private isInitialized = false;
  private isRecording = false;
  private isConnected = false;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000; // 1 second
  private pingInterval: NodeJS.Timeout | null = null;
  private status: RealtimeSTTStatus | null = null;

  constructor(config: RealtimeSTTConfig, callbacks: RealtimeSTTCallbacks) {
    this.config = config;
    this.callbacks = callbacks;
  }

  /**
   * Initialize the RealtimeSTT service
   */
  async initialize(): Promise<boolean> {
    try {
      console.log('🎤 Initializing RealtimeSTT service...');
      
      // Connect to WebSocket
      const success = await this.connect();
      if (!success) {
        console.error('❌ Failed to connect to RealtimeSTT service');
        return false;
      }

      this.isInitialized = true;
      console.log('✅ RealtimeSTT service initialized successfully');
      return true;
    } catch (error) {
      console.error('❌ Error initializing RealtimeSTT service:', error);
      this.callbacks.onError?.(`Initialization failed: ${error}`);
      return false;
    }
  }

  /**
   * Connect to the RealtimeSTT WebSocket
   */
  private async connect(): Promise<boolean> {
    return new Promise((resolve) => {
      try {
        const wsUrl = `${this.getWebSocketUrl()}/realtime-stt/stream`;
        console.log('🔌 Connecting to RealtimeSTT WebSocket:', wsUrl);
        
        this.ws = new WebSocket(wsUrl);
        
        this.ws.onopen = () => {
          console.log('✅ RealtimeSTT WebSocket connected');
          this.isConnected = true;
          this.reconnectAttempts = 0;
          this.startPingInterval();
          resolve(true);
        };

        this.ws.onmessage = (event) => {
          this.handleMessage(event);
        };

        this.ws.onclose = (event) => {
          console.log('🔌 RealtimeSTT WebSocket closed:', event.code, event.reason);
          this.isConnected = false;
          this.stopPingInterval();
          
          // Attempt to reconnect if not intentionally closed
          if (event.code !== 1000 && this.reconnectAttempts < this.maxReconnectAttempts) {
            this.attemptReconnect();
          }
        };

        this.ws.onerror = (error) => {
          console.error('❌ RealtimeSTT WebSocket error:', error);
          this.callbacks.onError?.(`WebSocket error: ${error}`);
          resolve(false);
        };

        // Timeout after 10 seconds
        setTimeout(() => {
          if (!this.isConnected) {
            console.error('❌ RealtimeSTT WebSocket connection timeout');
            this.ws?.close();
            resolve(false);
          }
        }, 10000);

      } catch (error) {
        console.error('❌ Error connecting to RealtimeSTT WebSocket:', error);
        resolve(false);
      }
    });
  }

  /**
   * Handle incoming WebSocket messages
   */
  private handleMessage(event: MessageEvent): void {
    try {
      const data = JSON.parse(event.data);
      
      switch (data.type) {
        case 'realtime_transcription':
          this.callbacks.onRealtimeTranscription?.(
            data.text,
            data.confidence,
            data.is_final
          );
          break;
          
        case 'final_transcription':
          this.callbacks.onFinalTranscription?.(
            data.text,
            data.confidence
          );
          break;
          
        case 'state_change':
          this.isRecording = data.is_recording;
          this.callbacks.onStateChange?.(data.is_recording);
          break;
          
        case 'language_detected':
          this.callbacks.onLanguageDetected?.(data.language);
          break;
          
        case 'status':
          this.status = data.status;
          break;
          
        case 'error':
          console.error('❌ RealtimeSTT error:', data.message);
          this.callbacks.onError?.(data.message);
          break;
          
        case 'pong':
          // Ping response received
          break;
          
        default:
          console.log('📨 Unknown RealtimeSTT message type:', data.type);
      }
    } catch (error) {
      console.error('❌ Error parsing RealtimeSTT message:', error);
    }
  }

  /**
   * Start the RealtimeSTT service
   */
  async start(): Promise<boolean> {
    try {
      if (!this.isConnected || !this.ws) {
        console.error('❌ RealtimeSTT not connected, cannot start');
        return false;
      }

      console.log('🎤 Starting RealtimeSTT service...');
      
      // Send start command
      this.sendMessage({
        type: 'start'
      });

      return true;
    } catch (error) {
      console.error('❌ Error starting RealtimeSTT service:', error);
      this.callbacks.onError?.(`Start failed: ${error}`);
      return false;
    }
  }

  /**
   * Stop the RealtimeSTT service
   */
  stop(): void {
    try {
      if (!this.isConnected || !this.ws) {
        console.log('⚠️ RealtimeSTT not connected, cannot stop');
        return;
      }

      console.log('🛑 Stopping RealtimeSTT service...');
      
      // Send stop command
      this.sendMessage({
        type: 'stop'
      });

      this.isRecording = false;
    } catch (error) {
      console.error('❌ Error stopping RealtimeSTT service:', error);
    }
  }

  /**
   * Update configuration
   */
  updateConfig(newConfig: Partial<RealtimeSTTConfig>): void {
    try {
      this.config = { ...this.config, ...newConfig };
      
      if (this.isConnected && this.ws) {
        this.sendMessage({
          type: 'update_config',
          ...newConfig
        });
      }
      
      console.log('⚙️ RealtimeSTT configuration updated');
    } catch (error) {
      console.error('❌ Error updating RealtimeSTT configuration:', error);
    }
  }

  /**
   * Get current status
   */
  async getStatus(): Promise<RealtimeSTTStatus | null> {
    try {
      if (this.isConnected && this.ws) {
        this.sendMessage({
          type: 'get_status'
        });
      }
      return this.status;
    } catch (error) {
      console.error('❌ Error getting RealtimeSTT status:', error);
      return null;
    }
  }

  /**
   * Send message to WebSocket
   */
  private sendMessage(message: any): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(message));
    } else {
      console.warn('⚠️ RealtimeSTT WebSocket not connected, cannot send message');
    }
  }

  /**
   * Start ping interval to keep connection alive
   */
  private startPingInterval(): void {
    this.pingInterval = setInterval(() => {
      if (this.isConnected && this.ws) {
        this.sendMessage({ type: 'ping' });
      }
    }, 30000); // Ping every 30 seconds
  }

  /**
   * Stop ping interval
   */
  private stopPingInterval(): void {
    if (this.pingInterval) {
      clearInterval(this.pingInterval);
      this.pingInterval = null;
    }
  }

  /**
   * Attempt to reconnect
   */
  private async attemptReconnect(): Promise<void> {
    this.reconnectAttempts++;
    console.log(`🔄 Attempting to reconnect RealtimeSTT (${this.reconnectAttempts}/${this.maxReconnectAttempts})...`);
    
    setTimeout(async () => {
      const success = await this.connect();
      if (success) {
        console.log('✅ RealtimeSTT reconnected successfully');
        // Restart if it was recording before
        if (this.isRecording) {
          await this.start();
        }
      } else {
        console.error('❌ RealtimeSTT reconnection failed');
      }
    }, this.reconnectDelay * this.reconnectAttempts);
  }

  /**
   * Get WebSocket URL based on environment
   */
  private getWebSocketUrl(): string {
    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
    const host = window.location.hostname;
    const port = process.env.NODE_ENV === 'production' ? '' : ':8000';
    return `${protocol}//${host}${port}`;
  }

  /**
   * Destroy the service and clean up resources
   */
  destroy(): void {
    try {
      console.log('🧹 Destroying RealtimeSTT service...');
      
      this.stop();
      this.stopPingInterval();
      
      if (this.ws) {
        this.ws.close(1000, 'Service destroyed');
        this.ws = null;
      }
      
      this.isInitialized = false;
      this.isRecording = false;
      this.isConnected = false;
      
      console.log('✅ RealtimeSTT service destroyed');
    } catch (error) {
      console.error('❌ Error destroying RealtimeSTT service:', error);
    }
  }
}

// Default configuration
export const defaultRealtimeSTTConfig: RealtimeSTTConfig = {
  model: 'base.en',
  language: 'en',
  useMicrophone: true,
  postSpeechSilenceDuration: 1.0,
  spinner: false,
  enableRealtimeTranscription: true,
  enableFinalTranscription: true,
  sampleRate: 16000,
  chunkSize: 1024
};

// Default callbacks
export const defaultRealtimeSTTCallbacks: RealtimeSTTCallbacks = {
  onRealtimeTranscription: (text: string, confidence: number, isFinal: boolean) => {
    console.log('🎤 Realtime transcription:', text, `(${confidence.toFixed(2)})`);
  },
  onFinalTranscription: (text: string, confidence: number) => {
    console.log('✅ Final transcription:', text, `(${confidence.toFixed(2)})`);
  },
  onError: (error: string) => {
    console.error('❌ RealtimeSTT error:', error);
  },
  onStateChange: (isRecording: boolean) => {
    console.log('🔄 RealtimeSTT state changed:', isRecording ? 'Recording' : 'Stopped');
  },
  onLanguageDetected: (language: string) => {
    console.log('🌍 Language detected:', language);
  }
};
